maintainer_guide.md

```diff
@@ -0,0 +1,80 @@
+      undocumented public interfaces, we must still consider backwards
+      compatibility when making changes to such potentially in-use code.
+    * For C++, code is private. If a change breaks fbcode, fix fbcode or revert
+      the change. We should be careful about models running in prod and relying
+      on torchvision ops. FYI PyTorch as BC API checks for its ops.
+* The test folder is not importable and is **private.** Even fbcode projects
+  should not rely on it. 
+* The references folder serves as documentation and is treated as such: it’s
+  “private” in the sense that there are no BC guarantees here. Breaking changes
+  are possible, but we should make sure that the tutorials are still running
+  properly, and that their intended narrative is preserved (by e.g. checking
+  outputs, etc.).
+* The rest of the folders (build, android, ios) are private and have no BC
+  breaking guarantees.
+
+### Should this attribute be public? Should this function be private?
+
+When designing an API it’s not always obvious what should be exposed as public,
+and what should be kept as a private implementation detail. The following
+guidelines can be useful:
+
+* Functional consistency throughout the library is a top priority, for users and
+  developers’ sake. In doubt and unless it’s clearly wrong, expose what other
+  similar classes expose.
+* Think really hard about the users and their use-cases, and try to expose what
+  they would need to address those use-cases. Aggressively keep everything else
+  private. Remember that the “private -> public” direction is way smoother than
+  the “public -> private” one: in doubt, keep it private.
+* When thinking about use-cases, the general API motto applies: make what’s
+  simple and common easy, and make what’s complex possible (80% / 20% rule).
+  There might be a ~1% left that’s not addressed: that’s OK. Also, make what’s
+  wrong very hard, if not impossible.
+* For models, things are more subtle as we still need a way to properly enable
+  model surgery. TODO: expand on this
+
+As a good practice, always create new files and even classes with a leading
+underscore in their name. This way, everything is private by default and the
+only public surface is explicitly present in an `__init__` file.
+
+### Should we deprecate X, or can we make a BC-breaking change?
+
+When a public-facing feature needs to change, we try to deprecate, but sometimes
+a BC-breaking change is possible or even warranted. Typically in the case of
+bugs, a failure is better than incorrect results. To decide whether to deprecate
+or to introduce a breaking change:
+
+1. Check whether X does something purely internal VS something that a reasonable
+   user might be interested in.
+2. Look for usages on Github for usages of X. This is often difficult due to the
+   forks, copy-pastes, and name conflicts.
+3. Look on FBcode where we can get a cleaner (but also biased) signal of the
```

Comment:
References to FBcode feel a bit strange here. Not all contributors can do this, so this probably needs to be rephrased to cover these cases (for example "sync with a FB contributor to do this for you". Quite common advice given also on PyTorch core).

Reply:
I added

> if you're not a fb employee, ask someone to do that for you.

But just to clarify, the audience of this document is the torchvision maintainers (fb employees + quantsight), not the general pool of contributors - although they might find some relevant info here too. fb and quantsight employees all know what fbcode is, which is why we're using the term "fbcode" as well above without further details

Reply:
Quantisight engineers wouldn't have access to FBcode either, so they still need to pair with a FB employee to do such checks.

Perhaps then the guide should leave in a public Google Docs (for transparency reasons) but mainly shared across the torchvision maintainers? Adding it inside the repo if it does not target the general pool of contributors might be confusing. Thoughts?

Reply:
It's very common for projects to include maintainers-only info in their repo / docs:

- https://scikit-learn.org/stable/developers/maintainer.html
- https://pandas.pydata.org/pandas-docs/stable/development/maintaining.html

There's usually no reason not to do so, and it help build trust / transparency. It's also easier to have everything at the same place.

Reply:
There's some subtle difference in whether we think of this as "an internal doc, posted publicly" or "a public explanation of our process"

I'm onboard with being transparent. If we're being transparent we can use language that's accessible to the readers to whom we're being transparent. E.g. if it's an internal only doc then terms like `fbcode` which reference a particular file path make sense. If we're being transparent so external contributors can see and understand our process, it seems slightly nicer to write out things like "facebook internal code" instead of referencing the directory name or internal abbreviations.

In some cases a statement like "We should/will/must check for..." might feel more inclusive than an imperative command directed at the reader ("Check for...") so that there's less of an implication that the reader is a FB employee.

though, I do notice examples of the imperative form in the referenced maintainer guides.

Reply:
> whether we think of this as "an internal doc, posted publicly" or "a public explanation of our process"

I think of it as both, with a stronger component on the first definition. It *may* be useful as a public explanation of our process in some rare cases, but it's not the main purpose.

Should non-maintainers read this document, I think it's fair to assume they would understand what "fbcode" means in this context, the name is fairly self-explanatory.

In pytorch core, there are many explicit references to fbcode:

```
git grep -i fbcode | wc -l
127
```

I'm happy to edit the document if you really insist on removing internal terms and language, but in all honesty, I don't think it will matter :)